### DISCUSSION TITLE
Ocaml-lsp can't find the generated OCaml code (e.g. `_build/default/src/lib/ast.ml`)

### DISCUSSION THREAD

**Discussion by Timmmm** [rems-project/sail#520] [Q&A]
I've been doing some work on Sail using VSCode with ocaml-lsp, which is surprisingly good, *except* that it can't find any of the code that's generated at compile time. In particular `_build/default/src/lib/ast.ml`, which is used all over the place.

I can't find any "additional include directories" style options in ocaml-lsp , which is what you'd typically do for C++, Python, etc. Just wondering if you have any idea how to fix this.

**Comment by Alasdair** [comment]
Even though it can't find the code (for go to definition), it should still work for completion/inline type annotations I think?

**Reply by Timmmm** [reply]
Yeah it still seems to show types on hover... Could this be because it only finds a compiled version of that file or something? Maybe [this](https://discuss.ocaml.org/t/merlin-goto-definition-inside-installed-libraries/2788/4)?

**Reply by bacam** [reply]
I think you're right about it finding the compiled version (as it's in the same place as all of the other compiled files).  Poking at it a little, it's the underlying "merlin" tool that can't find the source file, so you can see the same effect in (e.g.) emacs on Linux too.  The part I don't understand is why - the dune build tool knows about the file, and it provides information to merlin.

**Reply by Timmmm** [reply]
Hmm yeah I think it's treating `ast.ml` as an installed library: https://github.com/ocaml/dune/issues/6397

The workaround seems to be to use manually written `.merlin` files, but Dune doesn't generate them any more, and `dune ocaml dump-dot-merlin` / `dune ocaml merlin dump-config` seem to be broken for me (`No merlin configuration found.`).

I might ask on StackOverflow or maybe the Merlin repo.

**Reply by bacam** [reply]
It's not quite that; it's that dune tells it where to find the compiled files, but not the source.  Perhaps that's because alongside the generated files there are also copies of the plain source files, and it doesn't want merlin to refer editors to those to prevent you from editing the wrong one.

However, you might be able to use a manually altered `.merlin` file, you just need to give `dump-dot-merlin` at path: `dune ocaml dump-dot-merlin src/lib/`

**Reply by bacam** [reply]
(Which also confirms that it's dune that's not telling merlin where to look.)

**Reply by Timmmm** [reply]
> you just need to give dump-dot-merlin at path

Ah would be nice if the docs mentioned that! So that does give some output and `dune ocaml merlin dump-config src/lib` gives similar output.

I tried putting the `dump-dot-merlin` output in `.merlin`, and it's definitely reading it (tested via the age-old "add sdkfgjhgdsfkgjhd and see if it complains" method). I then add this to the VSCode config:

```
    "ocaml.server.args": [
        "--fallback-read-dot-merlin"
    ],
```

But after that it basically fails to resolve anything, including stuff that's in the same directory, e.g. `open Rewriter`. Hmm.
